File: list5/prolog_code/funs.pl
splitlist([], [], []).
splitlist([X], [X], []).
splitlist([X, Y | Rest], [X | Left], [Y | Right]) :- 
	splitlist(Rest, Left, Right).

merge([], Right, Right).
merge(Left, [], Left).
merge([L | Left], [R | Right], [L | Merged]) :-
	L =< R,
	merge(Left, [R | Right], Merged).
merge([L | Left], [R | Right], [R | Merged]) :-
	L > R,
	merge([L | Left], Right, Merged).

mergesort([], []).
mergesort([X], [X]).
mergesort(List, Sorted) :-
	splitlist(List, Left, Right),
	mergesort(Left, SortedLeft),
	mergesort(Right, SortedRight),
	merge(SortedLeft, SortedRight, Sorted).

de(A, 0, 1, 0, A).
de(A, B, X, Y, Z) :-
	B > 0,
	R is A mod B,
	Q is A // B,
	de(B, R, X1, Y1, Z),
	X is Y1,
	Y is X1 - Q * Y1.

pfactors(1, _, []).
pfactors(N, F, [F | Factors]) :-
	N > 1,
	R is N // F,
	N =:= R * F,
	pfactors(R, F, Factors).
pfactors(N, F, Factors) :-
	N > 1,
	R is N // F,
	N =\= R * F,
	next_factor(F, F1),
	pfactors(N, F1, Factors).

next_factor(2, 3).
next_factor(F, NextF) :-
    F > 2,
    NextF is F + 2.

primefactors(N, Factors) :-
	pfactors(N, 2, Factors).

gcd(A, 0, A).
gcd(A, B, C) :-
	B > 0,
	R is A mod B,
	gcd(B, R, C).

totient(1, 1).
totient(N, T) :-
	N > 1,
	findall(X, (between(1, N, X), gcd(N, X, 1)), Relatives),
	length(Relatives, T).

isprime(2).
isprime(3).
isprime(N) :-
    N > 3,
    N mod 2 =\= 0,
    \+ hasfactor(N, 3).

hasfactor(N, F) :-
    N mod F =:= 0.
hasfactor(N, F) :-
    F * F < N,
    F2 is F + 2,
    hasfactor(N, F2).

primes(N, P) :-
    findall(X, (between(2, N, X), isprime(X)), P).


:- mergesort([9, 8, 7, 6, 5, 4, 3, 2, 1], Y), write(Y), nl.
:- de(27, 15, X, Y, Z), write(X), nl, write(Y), nl, write(Z), nl.
:- primefactors(123, X), write(X), nl.
:- totient(123, X), write(X), nl.
:- primes(123, X), write(X), nl.

File: list5/haskell_code/funs.hs
import Data.List (nub)

binomial :: Integer -> Integer -> Integer
binomial n k | k == 0 = 1
             | k == n = 1
             | otherwise = binomial (n - 1) k + binomial (n - 1) (k - 1)

binomial2 :: Integer -> Integer -> Integer
binomial2 n k = (pascal_triangle !! fromIntegral(n) !! fromIntegral(k))
  where pascal_triangle = [[pascal i j | j <- [0..i]] | i <- [0..]]
        pascal _ 0 = 1
        pascal x y | x == y = 1
                   | otherwise = pascal_triangle !! (x - 1) !! (y - 1) + pascal_triangle !! (x - 1) !! y

mergesort :: Ord a => [a] -> [a]
mergesort [] = []
mergesort [x] = [x]
mergesort lst = merge (mergesort left) (mergesort right)
  where (left, right) = splitAt (length lst `div` 2) lst
        merge [] r = r
        merge l [] = l
        merge (l:ltail) (r:rtail) | l <= r = l : merge ltail (r:rtail)
                                  | otherwise = r : merge (l:ltail) rtail

de :: Integral a => a -> a -> (a, a, a)
de a 0 = (1, 0, a)
de a b =
  let (q, r) = a `quotRem` b
      (x, y, z) = de b r
  in (y, x - q * y, z)

prime_factors :: Integer -> [Integer]
prime_factors n = factor n 2
  where factor 1 _ = []
        factor n p | n `mod` p == 0 = p : factor (n `div` p) p
                   | otherwise = factor n (p + 1)

totient :: Integer -> Integer
totient n = fromIntegral $ length [x | x <- [1..n], gcd n x == 1]

totient2 :: Integer -> Integer
totient2 n = product [product $ subtract_one $ [x | x <- factors, x == p] | p <- unique_factors]
  where factors = prime_factors n
        unique_factors = nub factors
        subtract_one [] = []
        subtract_one (hd:tl) = (hd - 1):tl

primes :: Integer -> [Integer]
primes n = sieve [2..n]
  where sieve (p:lst) = p : sieve [x | x <- lst, x `mod` p /= 0]
        sieve [] = []

main :: IO ()
main = do {
    putStrLn $ show (binomial 7 3);
    putStrLn $ show (binomial2 7 3);
    putStrLn $ show (mergesort [9,8..1]);
    putStrLn $ show (de 27 15);
    putStrLn $ show (prime_factors 123);
    putStrLn $ show (totient 123);
    putStrLn $ show (totient2 123);
    putStrLn $ show (primes 123);
}
